packages/cli/lib/DevServerManager.js

```diff
@@ -124,6 +125,16 @@ class DevServerManager {
     }
   }
 
+  updateConfigFile({ filepath, config }) {
+    if (this.started) {
+      this.iterateDevServers(async serverInterface => {
+        if (serverInterface.updateConfigFile) {
+          await serverInterface.updateConfigFile(filepath, config);
```

Comment:
Hi @camden11, I'm trying to familiarize myself with these lifecycle methods. What's the difference between `fileChange` and `updateConfigFile`?

Reply:
Unfortunately to my knowledge there's no documentation (cc @brandenrodgers) but agreed that that's definitely something we should work on, especially if other teams are to develop their own dev servers.

As it stands now:
* `fileChange` is called by the CLI any time a file _other_ than an app config is changed. If an app config file is changed, the CLI assumes the dev server can't handle that change and prompts the user to upload their project.
* If that upload is successful, the CLI informs the dev server of the changed app config file via `updateConfigFile`

* If a card config file (or any other file the dev server doesn't want to handle or knows upload is required for) is changed, the CLI will still call `fileChange`. Within `fileChange`, the dev server will recognize the file as one that requires an upload, and call the `onUploadRequired` callback passed from the CLI. 
* This triggers the upload prompt. If it's successful, the CLI will then call `updateConfigFile` to let the dev server know the config file was updated

It is admittedly starting to seem like this prompt functionality doesn't work super nicely with the way the dev server interface was originally set up. Might be worth re-thinking entirely, but for the time being wanted something that works within the current constraints.

As for the inconsistency between `setup` and `updateConfigFile`, the current `updateConfigFile` API was recommended by @joe-yeager. Can definitely update to make it more similar to `setup` if you both think that would make more sense

Reply:
Another +1 for for formalizing the contract, that is definitely a gap.

> I'm also curious if there's any documentation that might explain the lifecycle of a dev server. For example, it would be helpful to know when is setup called, the expectation on the dev server, the contract between the dev server and the dev mode manager

`setup` is called [here](https://github.com/HubSpot/hubspot-cli/pull/1076/files/b4baee91d124c89dd75ef4ad36de5b5e65b6981c#diff-004104a9aade0cabe581c4be51b61d8b43745b9772cc1d3afedc9c5b64eb3d00R82) and should only be called when `hs project dev` is initializing.  It can technically be called multiple times because the `DevServerManager` iterates the dev servers defined in [this map](https://github.com/HubSpot/hubspot-cli/blob/main/packages/cli/lib/DevServerManager.js#L32-L42) and calls the `setup` method based on the components.  We have [this check](https://git.hubteam.com/HubSpot/ui-extensibility/blob/master/public-packages/ui-extensions-dev-server/src/lib/DevModeInterface.ts#L145) to prevent the dev server from running for multiple components types because right now the only components are public apps and private apps which are not supported in the same project.

> I feel like there's some inconsistency between the setup and updateConfigFile interfaces. Setup provides a data structure about the components ([here](https://git.hubteam.com/HubSpot/ui-extensibility/blob/master/public-packages/ui-extensions-dev-server/src/lib/DevModeInterface.ts#L42)), whereas updateConfigFile drops one level down to the file

The reason I recommended this API was that we can use the `filepath` to loop over the [extensionsConfig](https://git.hubteam.com/HubSpot/ui-extensibility/blob/master/public-packages/ui-extensions-dev-server/src/lib/DevModeInterface.ts#L67) which is an array of the configs for all of the currently running extensions and has the [`extensionsConfigPath`](https://git.hubteam.com/HubSpot/ui-extensibility/blob/master/public-packages/ui-extensions-dev-server/src/lib/types.ts#L13) prop, which should be the absolute file path of the extension config.  So the [projectComponentMap](https://git.hubteam.com/HubSpot/ui-extensibility/blob/master/public-packages/ui-extensions-dev-server/src/lib/types.ts#L100) seemed like extraneous information

>  What's the difference between fileChange and updateConfigFile?

> It is admittedly starting to seem like this prompt functionality doesn't work super nicely with the way the dev server interface was originally set up

This question and statement bring up a great point.  I'm just riffing, but maybe we could update the `fileChange` method to add an additional 'context' arg that could contain the config files contents to simplify the contract, we would also need to update the logic so that method is called even for app config files so the dev server can update it's local copy of the app config.  So it's contract could be something along the lines of:

```typescript
function fileChange(filePath: string, __event: unknown, context?: { config?: AppConfig | ExtensionConfig}){}
```

That feels like it gets us what we need while removing some of the complications.

Reply:
Here's a [placeholder issue](https://git.hubteam.com/HubSpot/hubspot-cli-issues/issues/581) to document the contract so that doesn't get left behind

Reply:
After looking into this approach, @joe-yeager and I decided that it would be safer and more straightforward to instead add a `restart` method to the dev server that stops and restarts the dev server while preserving any relevant state from the previous session.

PR for that here: https://git.hubteam.com/HubSpot/ui-extensibility/pull/2306